### ISSUE TITLE
quotes

### ISSUE THREAD

**Issue by frivoal** [mozdevs/cssremedy#6]
https://github.com/mozdevs/cssremedy/blob/7a41665f3168a9719a404f9693a6caa30b2b502d/remedy.css#L81

See https://github.com/whatwg/html/issues/3636#issuecomment-459922404 for a better selector to apply quotation marks.

**Comment by jensimmons** [comment]
Yeah. I actually talked to @fantasai after writing that one simple line of code, and she pointed me to
https://html.spec.whatwg.org/multipage/rendering.html#quotes
as the official What Browsers Should Be Doing for Default Quotes.

I took that doc and rewrote it to be
https://github.com/mozdevs/cssremedy/blob/master/quotes.css
which led her and I to have a discussion, and for her to post that comment you linked to.

```
:where(:root:lang(en), [lang]:lang(en))) { curly quotes }
:where(:root:lang(fr), [lang]:lang(fr))) { guillemots}
:where(:root:lang(zh), [lang]:lang(zh))) { brackets }
q { quotes: inherit; }
```

looks cool, but `:where` isn't implemented enough yet for us to count on it. We need a way that will do it today. (We can do both, as progressive enhancement, but if the result is the same, might not bother).

**Comment by jensimmons** [comment]
@fantasai also opened this: https://github.com/whatwg/html/issues/4333

**Comment by jensimmons** [comment]
Here's a test of quotes on `<q>`:
http://software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cp%3E%3Cq%3ETest%3C%2Fq%3E%0A%3Cp%20lang%3Dfr%3E%3Cq%3ETest%3C%2Fq%3E%0A%3Cp%20lang%3Den%3E%3Cq%3ETest%3C%2Fq%3E%0A%3Cp%20lang%3Dja%3E%3Cq%3ETest%3C%2Fq%3E%0A%3Cp%20lang%3Dzh%3E%3Cq%3ETest%3C%2Fq%3E

**Comment by jensimmons** [comment]
Well, these results are sad.

Chrome and Safari (on Mac) fail to render default quotes as curly quotes. Firefox fails to render any of the languages in Hixie's test as anything but curly quotes. Which means one browser or another gets most of these wrong. 

<img width="1077" alt="live dom viewer 2019-02-04 17-29-46" src="https://user-images.githubusercontent.com/108474/52242285-42aea480-28a4-11e9-86d4-7e8043f0e63b.png">

I tried out more languages in [a bigger test](http://software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cp%3E%3Cq%3ETest%3C%2Fq%3E%0A%3Cp%20lang%3Dfr%3E%3Cq%3ETest%3C%2Fq%3E%0A%3Cp%20lang%3Den%3E%3Cq%3ETest%3C%2Fq%3E%0A%3Cp%20lang%3Dja%3E%3Cq%3ETest%3C%2Fq%3E%0A%3Cp%20lang%3Dzh%3E%3Cq%3ETest%3C%2Fq%3E). The results were also bad. Bad browsers.

**Comment by jensimmons** [comment]
Florian also made a test: https://jsbin.com/savatoqala/edit?html,css,output

**Comment by frivoal** [comment]
That test of mine is for a different selector than the one proposed by fantasai, which gives very similar results. Fantasai's is faster, but depends on :where, which isn't implemented anywhere, and ignores documents that use xml:lang instead of lang. Mine uses `:not()` with a complex selector as an argument, which is supported in Safari, but (as far as I know) not elsewhere.

**Comment by jensimmons** [comment]
Yeah, at this point there are two questions at hand:

1) Should we bother? It's clear browsers are not defaulting to the correct quote marks, especially for languages around the globe. It does take many lines of code to fix this. Some browser engineers are arguing it's not worth the extra kb in the UA stylesheet to do this properly. Do we want to spend the Kbs? 

2) If we do this, how? Which selector should we use? We can write this code much more efficiently using newer selectors. Using newer selectors will mean it'll only work in newer browsers. (Or none, even.) Progressive enhancement would normally dictate we do several versions — but that will be even more Kbs and even slower — so we should only pick one method. Which method we use will affect speed in rendering from the browser itself. 

@frivoal Could you list the options for which. selector pattern to use, along with the pro/con & support for each? (Or some of that if that's too much work). That might help us understand what the options are. 

Even if we don't fix this in a "main" stylesheet" we might want to provide this as a custom build option. Include quotes or not. Or even include support for X, Y, Z languages (and not the rest).

**Comment by tigt** [comment]
I think `:lang()` selector already considers ancestors, so it may be more correct and robust to write the selectors more like:

```css
q:lang(en) { curly quotes }
q:lang(fr) { guillemots}
q:lang(zh) { brackets }
```

I believe it was invented because correctly selecting elements by language is nigh-impossible to do with `[lang|="foo"]` and descendant selectors, due to nesting.

**Comment by tigt** [comment]
Scratch that; the unusual selectors are due to `q[lang]` specifically: [whatwg/html#4333: Flatten selectors used for quotes?](https://github.com/whatwg/html/issues/4333)

It looks like there’s a lot to think about in this corner of the spec! Changes to the default suggested rendering are discussed in [whatwg/html#3636: Default styling for q element](https://github.com/whatwg/html/issues/3636). I particularly like @fantasai’s suggestion:

```css
@namespace xml "http://www.w3.org/XML/1998/namespace";
:where(:root:lang(en), [lang]:lang(en), [xml|lang]:lang(en))) { curly quotes }
:where(:root:lang(fr), [lang]:lang(fr), [xml|lang]:lang(fr))) { guillemots}
:where(:root:lang(zh), [lang]:lang(zh), [xml|lang]:lang(fr))) { brackets }
q { quotes: inherit; }
```

**Comment by jensimmons** [comment]
Engineers at Mozilla have been working on improving the Q quotes in Firefox lately. I should find the issue and surface where they are at now....

Also, it's occurred to me that my initial goal above ^ was to create a stylesheet for Q that would have _all_ the right quotes, for every language, and then make that stylesheet as compact and compressed as possible. 

The original we grabbed from the HTML spec is not at all tightened up. It's very verbose. 

The good thing about the verbose version is that a developer can pick through it and find just the code for the languages their site will contain. And then only ship that part.

Maybe we make a separate quotes stylesheet that has a few parts to it:
1. explain the problem
2. list all the options, commented out
3. provide the compressed & tightened version, that can just be shipped all of that... without review, not commented out.

Although that sort of doesn't make sense — because that whole file would be many more KBs than just doing part 3. Or part 2.  With all the comments, especially, it'd be the most verbose.... and the sort of point was to try not to bloat the whole project with KBs of stuff for quotes..... something to think about.
